feat: tuyau article (#26)
adonisjs/adonisjs.com: content/blog_posts/articles/introducing-tuyau.md

---
summary: Announcing Tuyau, a new library in the AdonisJS ecosystem for having an E2E typesafe client-server communication.
---

We are super excited to announce the first release of [Tuyau](https://github.com/Julien-Sponsors/tuyau). Tuyau is a collection of libraries that provides you with a **type-safe request client**, **type-safe Interia Link helpers**, and **an Open-API spec generator**.

[![](https://static.julr.dev/tuyau.png)](https://tuyau.julr.dev)

**Website:** [https://tuyau.julr.dev](https://tuyau.julr.dev) \

> **Github Repo:** [Julien-R44/tuyau](https://github.com/Julien-R44/tuyau)

Reviewer: ```suggestion
**GitHub Repo:** [Julien-R44/tuyau](https://github.com/Julien-R44/tuyau)
```

## Type-safe request client

> Before we dig deeper into the usage of Tayau's request client, let's look at the problem we are trying to solve here.

Reviewer: ```suggestion
Before we dig deeper into the usage of Tuyau's request client, let's look at the problem we are trying to solve here.
```

```bash
node ace add @tuyau/core
```

Then, install the client in your frontend:

```bash
npm install @tuyau/client
```

Now, you’ll need to generate the types. The `@tuyau/core` package exposes a command, `node ace tuyau:generate`, which will create a `.adonisjs/api.ts` file at the root of your AdonisJS project. This file contains all the types necessary for the client to function. This command does **not** run automatically at the moment. You will need to run it manually after certain changes in your AdonisJS project, such as:

- Adding a new route/controller to your project
- Adding a `request.validateUsing` call in your controller method

Other than that, you won’t need to run this command frequently. For example, if you update the controller method's return type or the Vine schema, you **don’t need** to run the command again.

Once done, you can import and configure the client in your front-end as follows:

```ts
/// <reference path="../../adonisrc.ts" />

import { createTuyau } from '@tuyau/client'
import { api } from '@your-monorepo/server/.adonisjs/api'

export const tuyau = createTuyau({
  api,
  baseUrl: 'http://localhost:3333',
})
```

> You’ll need to configure your mono repo so that your front end depends on your backend when importing the `.adonisjs/api.ts` file. More information [here](https://tuyau.julr.dev/docs/installation#sharing-the-api-definition).

Reviewer: ```suggestion
You’ll need to configure your monorepo so that your front end depends on your backend when importing the `.adonisjs/api.ts` file. More information [here](https://tuyau.julr.dev/docs/installation#sharing-the-api-definition).
```
